src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -12,202 +16,340 @@ namespace System.Buffers
+            if (startObject != null)
+            {
+                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
+            }
+            else
+            {
+                type = SequenceType.Empty;
+            }
 
             int length = endIndex - startIndex;
-            object startSegment = start.GetObject();
-            object endSegment = end.GetObject();
+            object endObject = end.GetObject();
 
-            if (type != SequenceType.IMemoryList && startSegment != endSegment)
+            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
                 ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
 
-            if (type == SequenceType.IMemoryList)
+            if (type == SequenceType.MultiSegment)
             {
-                IMemoryList<T> startMemoryList = (IMemoryList<T>)startSegment;
-                if (startSegment != endSegment)
+                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
+                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
+
+                if (startSegment != endObject)
                 {
-                    next = GetBufferCrossSegment(startIndex, end, ref startMemoryList, ref length);
+                    next = GetBufferCrossSegment(startIndex, endObject, ref startSegment, ref length);
                 }
 
-                data = startMemoryList.Memory.Slice(startIndex, length);
+                data = startSegment.Memory;
             }
             else if (type == SequenceType.Array)
             {
-                data = new ReadOnlyMemory<T>((T[])startSegment, startIndex, length);
+                Debug.Assert(startObject is T[]);
+
+                data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
             }
             else if (type == SequenceType.OwnedMemory)
             {
-                data = ((OwnedMemory<T>)startSegment).Memory.Slice(startIndex, length);
+                Debug.Assert(startObject is OwnedMemory<T>);
+
+                data = (Unsafe.As<OwnedMemory<T>>(startObject)).Memory;
             }
             else if (typeof(T) == typeof(char) && type == SequenceType.String)
             {
-                data = (ReadOnlyMemory<T>)(object)((string)startSegment).AsMemory(startIndex, length);
+                Debug.Assert(startObject is string);
+
+                data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
             }
             else
             {
                 data = default;
+                return false;
+            }
+
+            data = data.Slice(startIndex, length);
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
```

Comment:
Entire separate code path just for `First` :(.

Reply:
Without it `First` isn't really an optimized short path as it iterates for `Next` via `GetBufferCrossSegment`, but then throws away the result 😿